### PR TITLE
feat: add option to generate virtual keys for embedded collections without standalone links

### PR DESCRIPTION
**Pull request by usu**
This fixes the problem with empty embedded collections which don't come with an own standalone link. Such embedded collections can currently not be distinguished from normal array properties after normalization.

### To discuss
- naming of option `embeddedStandaloneListVirtualKeys`
- pattern for generation of virtual key
  - another fixed delimiter than `#`
  - hash
  - configurable delimiter
  - configurable key generation function
- add additional meta data in the virtual key, e.g. `_meta.embedded = true`? or is the missing self link already sufficient?
  the current proposal includes a `virtual` flag, which makes it easer to recognize, that the link is a generated/virtual link and not a real one

### ToDo
- [x] Adjust documentation

Also fixes a bug, if only `embeddedStandaloneListKey` is enabled (without virtual keys), but when the provided link is an array instead of a standalone link (see this test for further details: https://github.com/carlobeltrame/hal-json-normalizer/pull/64/files#diff-aead39d5dfee7fb3c9c4092038162577d24b4aa828af9213f89ca4f73c2cd5b4R748).

### CODE REVIEW

File: src/normalize.js [hunk 1]
```diff
@@ -1,8 +1,6 @@
-import camelCase from 'lodash/camelCase';
-import isArray from 'lodash/isArray';
-import cloneDeep from 'lodash/cloneDeep';
-import keys from 'lodash/keys';
-import merge from 'lodash/merge';
+import {
+  camelCase, isArray, cloneDeep, keys, merge, difference,
```
Comment (carlobeltrame) [inline src/normalize.js:2]: Is that import structure better for tree shaking? Or why the change?

File: src/normalize.js [hunk 2]
```diff
@@ -92,21 +94,65 @@ function extractAllLinks(json, uri, opts) {
...
+        if (uri in links && rel in links[uri] && isSingleLink(links[uri][rel])) {
+          ret[uri][rel] = links[uri][rel];
+          ret[links[uri][rel].href] = {
+            [opts.embeddedStandaloneListKey]: embedded[uri][rel],
+            [opts.metaKey]: { self: links[uri][rel].href },
+          };
+        } else if (opts.embeddedStandaloneListVirtualKeys
+          && rel !== opts.embeddedStandaloneListKey) {
+          // no standalone link provided --> generate virtual key
+          delete ret[uri][rel];
```
Comment (carlobeltrame) [inline src/normalize.js:131]: I don't fully understand why this `delete` is necessary. Can you explain? How can there be something stored there already? Is this just to mirror the short duplicated code segment on line 148?

File: test/normalize.spec.js [hunk 3]
```diff
@@ -703,6 +744,231 @@ describe('embedded', () => {
 
     expect(result).to.deep.equal(output);
   });
+
+  it('embedded list without standalone link ignores link array', () => {
```
Comment (carlobeltrame) [inline test/normalize.spec.js:748]: Not sure I understand how the link array is ignored in this test. Are you saying there could be different data in `_links.questions` and it should still yield the same output? The current test doesn't really prove that the link array is ignored in my opinion.

File: src/normalize.js [hunk 4]
```diff
@@ -92,21 +94,65 @@ function extractAllLinks(json, uri, opts) {
...
     });
+
+    // also check remaining link properties to search for a possible collection
+    // which is not embedded
+    if (opts.embeddedStandaloneListVirtualKeys) {
+      difference(
+        keys(links[uri]),
+        [...keys(embedded[uri]), opts.embeddedStandaloneListKey],
+      ).forEach((rel) => {
+        if (Array.isArray(links[uri][rel])) {
```
Comment (carlobeltrame) [inline src/normalize.js:147]: This if statement is missing test coverage for both the true and false case.

File: src/normalize.js [hunk 5]
```diff
@@ -92,21 +94,65 @@ function extractAllLinks(json, uri, opts) {
...
+        }
       }
     });
+
+    // also check remaining link properties to search for a possible collection
+    // which is not embedded
+    if (opts.embeddedStandaloneListVirtualKeys) {
+      difference(
+        keys(links[uri]),
+        [...keys(embedded[uri]), opts.embeddedStandaloneListKey],
```
Comment (carlobeltrame) [inline src/normalize.js:145]: Why do we need an exception for a link with this name here and on line 129? I thought we didn't even support a link rel called "items", or if the API really used this rel, one should use a different embeddedStandaloneListKey..?

**Review by carlobeltrame — CHANGES_REQUESTED** [review]
Thanks! I like the feature, and the tests helped a lot to get a gist of how it works in practice. Some remarks are below.

File: src/normalize.js (hunk 1 above)
Comment (usu) [reply src/normalize.js:2]: I thought they are equivalent, but seem they aren't:
https://www.azavea.com/blog/2019/03/07/lessons-on-tree-shaking-lodash/#:~:text=The%20lodash%2Des%20library%20supports,up%20in%20the%20app%20bundle.

Will revert.

File: src/normalize.js (hunk 2 above)
Comment (usu) [reply src/normalize.js:131]: Example from the test 'embedded list with link array generates virtual key':

When reaching this line, we have the following value for `ret` (populated in line 97/98):
```javascript
ret = {
  "http://example.com/posts/2620": {
    questions: [
      {
        href: "http://example.com/questions/295",
      },
    ],
  },
  "http://example.com/questions/295": {
    id: 295,
    text: "Why?",
    _meta: {
      expiresAt: 1513868982,
      self: "http://example.com/questions/295",
    },
  },
}
```

We know want to remove `ret['http://example.com/posts/2620']['questions']`, because we want to replace this array with our virtual link object. If we don't delete it here, we end up with this after merging the next line:
```
ret['http://example.com/posts/2620']['questions'] = 
{
  0: {
    href: 'http://example.com/questions/295',
  },
  href: 'http://example.com/posts/2620#questions',
  virtual: true,
}
```

File: src/normalize.js (hunk 5 above)
Comment (usu) [reply src/normalize.js:145]: Well, the API uses it for collections. API platform returns something like this:
```javascript
{
  "_links": {
    "self": {
      "href": "/activities"
    },
    "items": [
      {
        "href": "/activities/b3acd3cf5282"
      }
    ]
  },
  "totalItems": 5,
  "_embedded": {
    "items": [
      {
        "_links": {
          "self": {
            "href": "/activities/b3acd3cf5282"
          },
        },
        "title": "Neue Aktivität",
        "location": "test",
        "id": "b3acd3cf5282"
      }
    ]
  }
}
```

Without excluding items, this would extract #items from the collection and normalize to something like this:
```javascript
{
    "/activities": {
        "totalItems": 5,
        "items": {
            "href": "/activities#items",
            "virtual": true
        }
    },
    "/activities#items": {
        "items": [
            {
                "href": "/activities/b3acd3cf5282"
            }
        ]
    },
    "/activities/b3acd3cf5282": {
        "title": "Neue Aktivität",
        "location": "test",
        "id": "b3acd3cf5282"
    }
}
```

So normally (at least for us) `embeddedStandaloneListKey` matches the same key that is used to transmit collection items. However, the hal standard doesn't specify this AFAIK. If we want to make this more flexible, we could separate this 2 options and allow to configure an array of property names that is always excluded from generating virtual links.

File: src/normalize.js (hunk 4 above)
Comment (usu) [reply src/normalize.js:147]: Uh, I thought I had a test for this, but seems I added that one only in `hal-json-vuex`. Will add one.

File: test/normalize.spec.js (hunk 3 above)
Comment (usu) [reply test/normalize.spec.js:748]: Yeah, the title seems a bit misleading. I think a better wording would be "keep list embedded if data provided in links is not a standalone link".

What's happening here is the following: Before this PR, we would check if `_links.question`s exists (which it does) and use it as a standalone link (=[accessing it's href property](https://github.com/carlobeltrame/hal-json-normalizer/pull/64/files#diff-7f7c0e0d6a9d315240bed1f927d895f579dd0ff44188355588383be7ce26574dL104)). Which leads to an error, because `_links.questions` is an array and `_links.questions.href` doesn't exist.

Now I'm checking first, if `_links.questions` is really a standalone link (`isSingleLink`). And this is the test for it.

**Review by carlobeltrame — APPROVED** [review]
Thanks for the explanations!

I like the pattern for the generation of the key you chose, because the resulting URIs are still valid (at least as long as the API does not include an anchor in an URI already). If really needed, we could also easily make this more configurable in a future change.

While writing the documentation it occurred to me that we are missing `self` links in the normalized "virtual entities". I added this, for consistency's sake. The feature is opt-in, and using the `virtual: true` flag, one can easily distinguish a virtual from a normal self link.

Given these virtual self links, I'd also propose to rename it to `virtualSelfLinks`. I think that's a little easier to grasp than `embeddedStandaloneListVirtualKeys` (even though the longer one is a little more precise). The renaming will of course require a small change in https://github.com/ecamp/hal-json-vuex/pull/260

What do you think? Ready to merge?

**Comment by usu** [comment]
> While writing the documentation it occurred to me that we are missing `self` links in the normalized "virtual entities". I added this, for consistency's sake. The feature is opt-in, and using the `virtual: true` flag, one can easily distinguish a virtual from a normal self link.

An alternative idea I had earlier was to store the parent relation on the virtual links. Something like this:
```javascript
{
  'http://example.com/posts/2620#questions': {
          items: [
            {
              href: 'http://example.com/questions/295',
            },
          ],
          _meta: {
            self: 'http://example.com/posts/2620#questions',
            parentResource: 'http://example.com/posts/2620',
            parentProperty: 'questions'
          }
   }
}
```

This would also act as a detector, that the link is virtual. And it would also provide the owner of the embedded collection without the need to parse the self link. For `hal-json-vuex` we don't really need this at the moment, as we remember the parent link via separate logic. But could be an improvement later on for others or to simplify our own logic.

> Given these virtual self links, I'd also propose to rename it to `virtualSelfLinks`. I think that's a little easier to grasp than `embeddedStandaloneListVirtualKeys` (even though the longer one is a little more precise). The renaming will of course require a small change in [ecamp/hal-json-vuex#260](https://github.com/ecamp/hal-json-vuex/pull/260)

Good for me.

> What do you think? Ready to merge?

Yes. From my perspective ready to merge as is.

**Comment by carlobeltrame** [comment]
Nice idea with the `parentResource` and `parentProperty`, I think that would allow hal-json-vuex to throw away the added `reload` field, right? Anyways, we can add that later if we think it's worth it, and keep the `virtual: true` flag for backwards compatibility. Merging and releasing this PR now.
